### ISSUE TITLE
[autograd] backward() should not be allowed for complex scalar output

### ISSUE THREAD

**Issue by kshitij12345** [pytorch/pytorch#92750]
### 🐛 Describe the bug

```python
import torch

def fn(x):
    return (x * 0.5j).sum()


x = torch.ones(1, dtype=torch.double, requires_grad=True)
o = fn(x)
o.backward() # Shouldn't work
print(x.grad)
```

Output:
```
tensor([0.], dtype=torch.float64)
```

`backward()`  without passing `gradients` argument should work only for real scalar Tensor.

cc: @lezcano 

### Versions

master

cc @ezyang @gchanan @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @anjali411 @dylanbespalko @mruberry

**Comment by lezcano** [comment]
We should fix this. Gradients are only defined for functions into the real numbers.
